File: app/src/main/java/com/github/premnirmal/ticker/network/StupidYahooWrapConverter.java
package com.github.premnirmal.ticker.network;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import retrofit.converter.ConversionException;
import retrofit.mime.TypedInput;

/**
 * Created by premnirmal on 12/21/14.
 */
class StupidYahooWrapConverter extends BaseConverter {

    private static final Pattern PATTERN_RESPONSE = Pattern.compile("YAHOO\\.Finance\\.SymbolSuggest\\.ssCallback\\((\\{.*?\\})\\)");

    @Override
    public Object fromBody(TypedInput body, Type type) throws
            ConversionException {
        try {
            final String bodyString = getString(body.in());
            final Matcher m = PATTERN_RESPONSE.matcher(bodyString);
            if (m.find()) {
                final Suggestions suggestions = gson.fromJson(m.group(1), Suggestions.class);
                return suggestions;
            }
            throw new RuntimeException("Invalid response");
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: app/src/main/kotlin/com/github/premnirmal/ticker/network/StupidYahooWrapConverter.kt
package com.github.premnirmal.ticker.network

import com.github.premnirmal.ticker.network.data.SuggestionsNet
import com.google.gson.Gson
import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.Converter
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException
import java.lang.reflect.Type
import java.util.regex.Pattern

/**
 * Created by premnirmal on 3/3/16.
 */
private class StupidYahooWrapConverter(gson: Gson) : BaseConverter<SuggestionsNet>(gson) {

  override fun convert(value: ResponseBody?): SuggestionsNet? {
    try {
      val bodyString = value?.let { getString(it.byteStream()) }
      val m = PATTERN_RESPONSE.matcher(bodyString)
      if (m.find()) {
        return gson.fromJson(m.group(1), SuggestionsNet::class.java)
      }
      throw error("Invalid response")
    } catch (e: IOException) {
      return null
    }
  }

  companion object {

    private val PATTERN_RESPONSE =
      Pattern.compile("YAHOO\\.Finance\\.SymbolSuggest\\.ssCallback\\((\\{.*?\\})\\)")
  }
}

internal class YahooSuggestionsConverterFactory(private val gson: Gson,
                                                private val gsonConverterFactory: GsonConverterFactory) :
    Converter.Factory() {
  override fun responseBodyConverter(
      type: Type?,
      annotations: Array<out Annotation>?,
      retrofit: Retrofit?
  ): Converter<ResponseBody, *> {
    return StupidYahooWrapConverter(gson)
  }

  override fun requestBodyConverter(
      type: Type,
      parameterAnnotations: Array<out Annotation>,
      methodAnnotations: Array<out Annotation>,
      retrofit: Retrofit
  ): Converter<*, RequestBody>? {
    return gsonConverterFactory.requestBodyConverter(
        type, parameterAnnotations,
        methodAnnotations, retrofit
    )
  }

  override fun stringConverter(
      type: Type,
      annotations: Array<out Annotation>,
      retrofit: Retrofit
  ): Converter<*, String>? {
    return gsonConverterFactory.stringConverter(type, annotations, retrofit)
  }
}